### PR TITLE
adding schema

### LINKED ISSUE #22: Skill level service concept rework & schema
## Description
Concept for the skill levels: They work like a "rank" in games, your skill levels increase when you do stuff and can reach max level 100.
Add this to our concept document

Define the graphql schema for the skill level service

## Acceptance Criteria

- [ ] Concept document has been updated
- [ ] schema is documented
- [ ] Mark fields or tell @myluki2000  which fields need to be resolved by the Mesh gateway

### PR DESCRIPTION
**Pull request by Keyshal-211**
## Description of changes
Please provide a short description of the change or if the change is sufficiently described in the GitHub issue, provide a link to the issue.

  
## How has this been tested?
Please describe the test strategy you followed.

- [ ] automated unit test
- [ ] automated integration test
- [ ] automated acceptance test
- [ ] manual, exploratory test

In case of manual test, please document the test well including a set of user instructions and prerequisites. Each including an action, it's result, and where appropriate a screenshot.
In case of automated tests, the test coverage should be reasonably high, especially on Service classes and other classes with complex logic.
    
## Checklist before requesting a review
- [ ] My code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My code fulfills all acceptance criteria
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] The test coverage (line and branch) is reasonably high, especially on Service classes and other classes with complex logic
- [ ] I have made corresponding changes to the documentation
- [ ] I have added explanation of architectural decision and rationales to [wiki/adr](https://github.com/IT-REX-Platform/wiki/tree/main/adr)
- [ ] I have updated the changes in the ticket description

## Checklist for reviewer
- [ ] The code works and does not throw errors
- [ ] The code is easy to understand and there are no confusing parts
- [ ] The code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [ ] The code change accomplishes what it is supposed to do
- [ ] I cannot think of any use case in which the code does not behave as intended
- [ ] The added and existing tests reasonably cover the code change
- [ ] I cannot think of any test cases, input or edge cases that should be tested in addition
- [ ] Description of the change is included in the documentation

### CODE REVIEW

File: src/main/resources/graphql/schema.graphqls [hunk 1]
```diff
@@ -1,16 +1,96 @@
...
+type Mutation {
+  """
+  Update the skill levels of the current user for all skill types (Bloom taxonomy) in a chapter of a course.
+  """
+  updateSkillLevels(courseId: UUID!, chapterId: UUID!, levels: SkillLevelInput!): SkillLevels!
+
+  """
+  Manually update the skill levels of a user for all skill types (Bloom taxonomy) in a chapter of a course.
+  """
+  manualUpdateSkillLevels(userId: UUID!, courseId: UUID!, chapterId: UUID!, levels: SkillLevelInput!): SkillLevels!
```
Comment (PaulBredl) [inline src/main/resources/graphql/schema.graphqls]: What is the difference to the query above?

File: src/main/resources/graphql/schema.graphqls [hunk 2]
```diff
@@ -1,16 +1,96 @@
...
+  wholeCourseTotalSkillPoints(courseId: UUID!): Int!
 }
 
-type Query {
-  templates: [Template]
+type Mutation {
+  """
+  Update the skill levels of the current user for all skill types (Bloom taxonomy) in a chapter of a course.
+  """
+  updateSkillLevels(courseId: UUID!, chapterId: UUID!, levels: SkillLevelInput!): SkillLevels!
```
Comment (PaulBredl) [inline src/main/resources/graphql/schema.graphqls]: I mean it kinda makes sense for testing stuff but the update of skill levels is done via dapr events. So this mutation (and the one below) should not be necessary

File: src/main/resources/graphql/schema.graphqls [hunk 3]
```diff
@@ -1,16 +1,96 @@
...
+  """
+  Manually update the skill levels of a user for all skill types (Bloom taxonomy) in a chapter of a course.
+  """
+  manualUpdateSkillLevels(userId: UUID!, courseId: UUID!, chapterId: UUID!, levels: SkillLevelInput!): SkillLevels!
+}
+
+"""
+Represents the skill levels for a user in all skill types (Bloom taxonomy) for a chapter of a course.
+"""
+type SkillLevels {
```
Comment (PaulBredl) [inline src/main/resources/graphql/schema.graphqls]: and where are the actual skill levels here? 😅 there is only the ids and a log, not the actual skill levels.

File: src/main/resources/graphql/schema.graphqls [hunk 4]
```diff
@@ -1,16 +1,96 @@
...
+  """
+  The skill level of the user for the skill type "Analyze".
+  """
+  analyze: Int!
+}
+
+"""
+Represents a log entry for a skill level change.
+"""
+type SkillLevelLog {
```
Comment (PaulBredl) [inline src/main/resources/graphql/schema.graphqls]: I would do it the same as in the reward service, i.e., include the old value, the new value, the difference, a timestamp, a reason and associated content ids.

File: src/main/resources/graphql/schema.graphqls [hunk 5]
```diff
@@ -1,16 +1,96 @@
...
+  """
+  The logs for skill level changes.
+  """
+  logs: [SkillLevelLog!]!
+}
+
+"""
+An object containing skill levels for all skill types (Bloom taxonomy) for a user in a chapter of a course.
+"""
+input SkillLevelInput {
```
Comment (PaulBredl) [inline src/main/resources/graphql/schema.graphqls]: If you choose to remove the mutations, these inputs are no longer required.

File: src/main/resources/graphql/schema.graphqls [hunk 6]
```diff
@@ -1,16 +1,96 @@
...
+  """
+  courseId: UUID!
+  """
+  The ID of the chapter.
+  """
+  chapterId: UUID!
+  """
+  The logs for skill level changes.
+  """
+  logs: [SkillLevelLog!]!
```
Comment (PaulBredl) [inline src/main/resources/graphql/schema.graphqls]: The log should be specific for each skill level (apply, understand, etc)

File: src/main/resources/graphql/schema.graphqls [hunk 7]
```diff
@@ -1,16 +1,96 @@
...
+
+  """
+  Get the skill levels of the specified user for all skill types (Bloom taxonomy) in a chapter of a course.
+  """
+  skillLevelsForUser(userId: UUID!, courseId: UUID!, chapterId: UUID!): SkillLevels!
 
-# TODO define data types, queries and mutations
-# this can be done in a separate files as long as they are in this folder and
-# end with .graphqls
+  """
```
Comment (PaulBredl) [inline src/main/resources/graphql/schema.graphqls:23]: What is the point of these following three queries? What even are the total skill points? Summed over all categories? And what would be the use case?

File: src/main/resources/graphql/schema.graphqls [hunk 8]
```diff
@@ -1,16 +1,96 @@
...
+type Query {
+  """
+  Get the skill levels of the current user for all skill types (Bloom taxonomy) in a chapter of a course.
+  """
+  userSkillLevels(courseId: UUID!, chapterId: UUID!): SkillLevels!
+
+  """
+  Get the skill levels of the specified user for all skill types (Bloom taxonomy) in a chapter of a course.
+  """
+  skillLevelsForUser(userId: UUID!, courseId: UUID!, chapterId: UUID!): SkillLevels!
```
Comment (PaulBredl) [inline src/main/resources/graphql/schema.graphqls]: We need those two queries also aggregated over the whole course, so userSkillLevels and skillLevelsForUser, without a chapter id. I would suggest to make the chapterId optional and note in the comment, that when the chapterId is `null`, then the skill levels for the course are returned.

**Review by PaulBredl — CHANGES_REQUESTED** [review]
A good start, but needs some work

File: src/main/resources/graphql/schema.graphqls [hunk 9]
```diff
@@ -1,16 +1,110 @@
...
+  """
+  ASSESSMENT_LEARNING,
+  """
+  The user repeated an already done assessment, resulting in an increase in skill levels.
+  """
+  ASSESSMENT_REPEATING,
+  """
+  The user used hints while attempting an assessment, resulting in a smaller increase in skill levels.
+  """
+  ASSESSMENT_WITH_HINTS,
```
Comment (PaulBredl) [inline src/main/resources/graphql/schema.graphqls]: Those last three reaons are an interesting idea! But what would happen if both hints were used and the correctness changes?

Comment (Keyshal-211) [reply src/main/resources/graphql/schema.graphqls]: By using a weighted approach, we can maintain a fair representation of the user's actual skill level while taking into account the influence of hints and the correctness of responses. 
Weighting the Factors: To handle the combination of hints used and correctness changes, we can apply weighting to each factor. For example, we could assign percentages to each factor to determine their contribution to the final skill level change.
scenario could be: 
1. If the user uses hints and responds correctly, the skill level increase might be, for example, 30% of what it would be without hints.
2. If the user uses hints and responds incorrectly, the skill level might increase by only 10% of what it would be without hints.
3. If the user doesn't use hints and responds correctly, the skill level would increase significantly, let's say 90% of what it would be without hints.
4. If the user doesn't use hints and responds incorrectly, the skill level might not increase or even decrease slightly.

is it good idea? what you think @PaulBredl

Comment (Keyshal-211) [reply src/main/resources/graphql/schema.graphqls]: because I created that reason to looked into concept that we have for skill level: 
Skill levels are increased... 
• when learning new assessments or repeating
already done assessments
• less if student used hints 
• less depending on
correctness 
• depending on the time (on
timed assessments)

File: src/main/resources/graphql/schema.graphqls [hunk 10]
```diff
@@ -1,16 +1,110 @@
...
+enum SkillLevelChangeReason {
+  """
+  The user failed in an assessment, resulting in a decrease in skill levels.
+  """
+  ASSESSMENT_FAILED,
+  """
+  The user learned from a new assessment or repeated an already done assessment,
+  resulting in an increase in skill levels.
+  """
+  ASSESSMENT_LEARNING,
```
Comment (PaulBredl) [inline src/main/resources/graphql/schema.graphqls]:
```suggestion
  ASSESSMENT_LEARNED,
```

File: src/main/resources/graphql/schema.graphqls [hunk 11]
```diff
@@ -1,16 +1,110 @@
...
+  """
+  The user repeated an already done assessment, resulting in an increase in skill levels.
+  """
+  ASSESSMENT_REPEATING,
```
Comment (PaulBredl) [inline src/main/resources/graphql/schema.graphqls]:
```suggestion
  ASSESSMENT_REVIEWED,
```

File: src/main/resources/graphql/schema.graphqls [hunk 12]
```diff
@@ -1,16 +1,98 @@
...
+  """
+  newValue: Int!
+  """
+  The reason why the skill level has changed.
+  """
+  reason: SkillLevelChangeReason!
+  """
+  The ids of the chapters that are associated with the change.
+  """
+  associatedChapterIds: [UUID!]!
```
Comment (PaulBredl) [inline src/main/resources/graphql/schema.graphqls]: Why chapters? Should be contents
```suggestion
  associatedContentIds: [UUID!]!
```

File: src/main/resources/graphql/schema.graphqls [hunk 13]
```diff
@@ -1,16 +1,98 @@
...
+  """
+  The new skill level.
+  """
+  newValue: Int!
+  """
+  The reason why the skill level has changed.
+  """
+  reason: SkillLevelChangeReason!
+  """
+  The ids of the chapters that are associated with the change.
```
Comment (PaulBredl) [inline src/main/resources/graphql/schema.graphqls]:
```suggestion
  The ids of the contents that are associated with the change.
```

**Review by PaulBredl — APPROVED** [review]
